tklock.c

```diff
@@ -294,6 +304,81 @@ static void disable_autorelock_policy(void)
 	return;
 }
 
+/**
+ * Enable/disable X11 input events (currently only for touchscreen)
+ *
+ * @param enable TRUE enable events, FALSE disable events
+ * @return TRUE on success, FALSE on failure
+ */
+static gboolean generic_x11_event_control(const gboolean enable)
+{
+	gboolean ret = FALSE;
+	Display* dpy = NULL;
+	XDeviceInfo *devinfo = NULL;
+	int ndev = 0;
+	unsigned char data = (unsigned char)enable;
+
+	dpy = XOpenDisplay(NULL);
+	if (dpy == NULL) {
+		dpy = XOpenDisplay(":0.0");
+	}
+
+	if (dpy == NULL) {
+		mce_log(LL_INFO, "%s: unable to open display", __func__);
+		return ret;
+	}
+
+	if (x11_atom_touchscreen == None)
+		x11_atom_touchscreen = XInternAtom(dpy, XI_TOUCHSCREEN, True);
+
+	if (x11_atom_device_enabled == None)
+		x11_atom_device_enabled = XInternAtom(dpy, "Device Enabled", False);
+
+	if ((x11_atom_touchscreen == None) || (x11_atom_device_enabled == None)) {
+		mce_log(LL_WARN, "%s: unable to obtain X11 Atoms", __func__);
+		goto done;
+	}
+
+	devinfo = XListInputDevices(dpy, &ndev);
+
+	for (int i = 0; i < ndev; i++) {
+		XDeviceInfo info = devinfo[i];
+
+		if (info.use != IsXExtensionPointer)
+			continue;
+		if (info.type != x11_atom_touchscreen)
+			continue;
+
+		if ((x11_atom_device_enabled_type == None) || (x11_atom_device_enabled_format == 0)) {
+			unsigned long ignore_bytes_after, ignore_nitems;
+			unsigned char *ignore_data = NULL;
+
+			if (XIGetProperty(dpy, info.id, x11_atom_device_enabled, 0, 0, False, AnyPropertyType, &x11_atom_device_enabled_type, &x11_atom_device_enabled_format, &ignore_nitems, &ignore_bytes_after, &ignore_data)) {
+				mce_log(LL_WARN, "%s: unable to obtain X11 Device Enabled property atom type", __func__);
+				goto done;
+			} else {
+				XFree(ignore_data);
+			}
+		}
+
+		XIChangeProperty(dpy, info.id, x11_atom_device_enabled,
+				x11_atom_device_enabled_type,
+				x11_atom_device_enabled_format, PropModeReplace, &data, 1);
+
+		ret = TRUE;
+	}
+
+done:
+	if (dpy != NULL) {
+		XCloseDisplay(dpy);
+	}
+	if (devinfo != NULL) {
```

Comment:
Tool late for that check here :), see https://mail.gnome.org/archives/commits-list/2016-October/msg04241.html. Please check for devinfo != NULL before entering the loop and call XFreeDeviceList() inside that check.

Reply:
You will want to fix that here then too: https://github.com/maemo-leste/hildon-desktop/blob/cursor_management/src/main.c#L405

Reply:
Also, I initialise `ndev` to zero, so this should be fine?

Reply:
Changed the code anyway, but I think me setting `ndev` to zero should have been enough.

Reply:
> You will want to fix that here then too: https://github.com/maemo-leste/hildon-desktop/blob/cursor_management/src/main.c#L405

For sure

Reply:
> Changed the code anyway, but I think me setting `ndev` to zero should have been enough.

With the lack of proper documentation on how those functions work, better safe than sorry.

Reply:
Sure, already fixed it. :)